### PR TITLE
Add Amsgrad optimizer

### PR DESCRIPTION
**Pull request by jinxing94**
Amsgrad optimizer is a variant of the Adam optimizer. It is proposed in the ICLR 2018 paper titled [on the convergence of Adam and beyond](https://openreview.net/pdf?id=ryQu7f-RZ) .  

The paper argues that the exponential moving average used in Adam optimizer may cause the model fail to converge to an optimal solution, even in some simple convex optimization settings.  Amsgrad is proposed to address the issue by endowing the algorithm with “long-term memory” of past gradients.

Both tensorflow and pytorch have added the amsgrad optimizer in their official APIs.  See the amsgrad option in the Adam APIs of tensorflow and pytorch for reference.

### CODE REVIEW

File: paddle/fluid/operators/optimizers/amsgrad_op.cc [hunk 1]
```diff
@@ -0,0 +1,165 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (shippingwang) [inline paddle/fluid/operators/optimizers/amsgrad_op.cc]: 2016->2019

File: paddle/fluid/operators/optimizers/amsgrad_op.cc [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                     "when not zero, if param row size is larger then "
+                     "min_row_size_to_use_multithread and "
+                     "inner_op_parallelism is larger then 0, sparse update "
+                     "will run in multithread mode")
+        .SetDefault(1000);
+
+    AddComment(R"DOC(
+Amsgrad Optimizer.
+
+This implements the Amsgrad optimizer from Section 2 of the Amsgrad
```
Comment (shippingwang) [inline paddle/fluid/operators/optimizers/amsgrad_op.cc]: same as Adam??

**Review by shippingwang — CHANGES_REQUESTED** [review]
*haven't finished the review yet

File: paddle/fluid/operators/optimizers/amsgrad_op.cc (hunk 1 above)
Comment (jinxing94) [reply paddle/fluid/operators/optimizers/amsgrad_op.cc]: Thanks for pointing this. I have changed it in a new commit.

File: paddle/fluid/operators/optimizers/amsgrad_op.cc (hunk 2 above)
Comment (jinxing94) [reply paddle/fluid/operators/optimizers/amsgrad_op.cc]: No. Thanks for pointing the typo.  I have updated it in a new commit.
